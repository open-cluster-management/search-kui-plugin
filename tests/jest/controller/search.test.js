/*******************************************************************************
* Licensed Materials - Property of IBM
* (c) Copyright IBM Corporation 2019. All Rights Reserved.
*
* Note to U.S. Government Users Restricted Rights:
* Use, duplication or disclosure restricted by GSA ADP Schedule
* Contract with IBM Corp.
*******************************************************************************/
'use strict'
const search = require('../../../dist/controller/search')
const pluginState = require('../../../dist/pluginState')

const args = {
  argv: ['search', 'kind:pod', 'name:logging-elk-filebeat-ds-62jxh'],
  command: ['search kind:pod name:logging-elk-filebeat-ds-62jxh']
}
const filters = ['cluster', 'kind', 'label', 'name', 'namespace', 'status', 'apiVersion', 'available', 'created']

describe('Search plugin state', () => {
  const spy = jest.spyOn(pluginState, 'getPluginState')
  const _ = pluginState.getPluginState()

  expect(spy).toBeCalled()

  it('should get the plugin state', () => {
    expect(pluginState.getPluginState()).toMatchSnapshot()
  })

  it('should set the plugin state to true', () => {
    pluginState.setPluginState('enabled', true)
    expect(pluginState.getPluginState().enabled).toBe(true)
    expect(pluginState.getPluginState().enabled).toMatchSnapshot()
  })

  it('should set the plugin state to false', () => {
    pluginState.setPluginState('enabled', false)
    expect(pluginState.getPluginState().enabled).toBe(false)
    expect(pluginState.getPluginState().enabled).toMatchSnapshot()
  })

  it('should return a default array of filter values', () => {
    expect(pluginState.getPluginState().default).toMatchSnapshot()
  })

  it('should set the searchSchema filter values', () => {
    pluginState.setPluginState('searchSchema', filters)
    expect(pluginState.getPluginState().searchSchema.length).not.toBe(0)
    expect(pluginState.getPluginState().searchSchema).toMatchSnapshot()
  })

  it('should set the error value', () => {
    pluginState.setPluginState('error', 'Error: Connection timeout')
    expect(pluginState.getPluginState().error).toMatchSnapshot()
  })

  it('should get an updated plugin state', () => {
    expect(pluginState.getPluginState()).toMatchSnapshot()
  })
})

describe('Is search available', () => {
  const spy = jest.spyOn(search, 'isSearchAvailable')
  const _ = search.isSearchAvailable()

  expect(spy).toBeCalled()

  it('should return if search is available', () => {
    expect(search.isSearchAvailable()).toMatchSnapshot()
  })
})

describe('Rendering search available', () => {
  const spy = jest.spyOn(search, 'renderSearchAvailable')
  const _  = search.renderSearchAvailable(true, undefined)

  expect(spy).toBeCalled()

  it('should render that search is installed', () => {
    pluginState.setPluginState('enabled', true)
    pluginState.setPluginState('error', undefined)
    expect(search.renderSearchAvailable(pluginState.getPluginState().enabled, pluginState.getPluginState().error)).toMatchSnapshot()
  })

  it('should render that search is installed, but is not available', () => {
    pluginState.setPluginState('enabled', true)
    pluginState.setPluginState('error', 'Error: Connection timeout')
    expect(search.renderSearchAvailable(pluginState.getPluginState().enabled, pluginState.getPluginState().error)).toMatchSnapshot()
  })

  it('should render that search is not installed', () => {
    pluginState.setPluginState('enabled', false)
    pluginState.setPluginState('error', undefined)
    expect(search.renderSearchAvailable(pluginState.getPluginState().enabled, pluginState.getPluginState().error)).toMatchSnapshot()
  })

  it('should render that there was an issue checking if search was installed', () => {
    pluginState.setPluginState('enabled', false)
    pluginState.setPluginState('error', 'Error: Connection timeout')
    expect(search.renderSearchAvailable(pluginState.getPluginState().enabled, pluginState.getPluginState().error)).toMatchSnapshot()
  })
})

describe('Search command', () => {
  const spy = jest.spyOn(search, 'doSearch')
  const _ = search.doSearch(args)

  expect(spy).toBeCalled()

  it('should call doSearch', () => {
    expect(search.doSearch(args)).toMatchSnapshot()
    expect(search.doSearch).toMatchSnapshot()
  })
})
