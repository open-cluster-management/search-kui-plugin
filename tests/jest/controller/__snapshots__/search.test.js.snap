// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Is search available should return if search is available 1`] = `false`;

exports[`Rendering search available should render that search is installed 1`] = `
<div
  class="is-search-available"
>
  <div>
    <p>
      The search service is installed and available.
    </p>
  </div>
</div>
`;

exports[`Rendering search available should render that search is installed, but is not available 1`] = `
<div
  class="is-search-available"
>
  <div>
    <p>
      <span
        class="oops"
      >
        The search service is installed, but the connection was not completed. Refresh the page to try connecting again. If the problem persists, contact your cluster administrator.
      </span>
    </p>
  </div>
</div>
`;

exports[`Rendering search available should render that search is not installed 1`] = `
<div
  class="is-search-available"
>
  <div>
    <p>
      The search service is not installed on this cluster. You must install the 
      <span
        class="install-details-link"
      >
        Cloud Pak for Multicloud Management
      </span>
       to use the search command.
    </p>
  </div>
</div>
`;

exports[`Rendering search available should render that there was an issue checking if search was installed 1`] = `
<div
  class="is-search-available"
>
  <div>
    <p>
      <span
        class="oops"
      >
        The search service is unavailable. If the problem persist, contact your cluster administrator.
      </span>
    </p>
  </div>
</div>
`;

exports[`Search command should call doSearch 1`] = `Promise {}`;

exports[`Search command should call doSearch 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "argv": Array [
          "search",
          "kind:pod",
          "name:logging-elk-filebeat-ds-62jxh",
        ],
        "command": Array [
          "search kind:pod name:logging-elk-filebeat-ds-62jxh",
        ],
      },
    ],
    Array [
      Object {
        "argv": Array [
          "search",
          "kind:pod",
          "name:logging-elk-filebeat-ds-62jxh",
        ],
        "command": Array [
          "search kind:pod name:logging-elk-filebeat-ds-62jxh",
        ],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`Search plugin state should get an updated plugin state 1`] = `
Object {
  "default": Array [
    "cluster",
    "kind",
    "label",
    "name",
    "namespace",
    "status",
  ],
  "enabled": false,
  "error": "Error: Connection timeout",
  "searchSchema": Array [
    "cluster",
    "kind",
    "label",
    "name",
    "namespace",
    "status",
    "apiVersion",
    "available",
    "created",
  ],
}
`;

exports[`Search plugin state should get the plugin state 1`] = `
Object {
  "default": Array [
    "cluster",
    "kind",
    "label",
    "name",
    "namespace",
    "status",
  ],
  "enabled": undefined,
  "error": undefined,
  "searchSchema": Array [],
}
`;

exports[`Search plugin state should return a default array of filter values 1`] = `
Array [
  "cluster",
  "kind",
  "label",
  "name",
  "namespace",
  "status",
]
`;

exports[`Search plugin state should set the error value 1`] = `"Error: Connection timeout"`;

exports[`Search plugin state should set the plugin state to false 1`] = `false`;

exports[`Search plugin state should set the plugin state to true 1`] = `true`;

exports[`Search plugin state should set the searchSchema filter values 1`] = `
Array [
  "cluster",
  "kind",
  "label",
  "name",
  "namespace",
  "status",
  "apiVersion",
  "available",
  "created",
]
`;
